@dalg24 https://github.com/kokkos/kokkos/pull/9139#discussion_r3380400775

All `get_kernel_func` in `core/src/Cuda/Kokkos_Cuda_KernelLaunch.hpp` but the one I've fixed seem to return this way.

The error I got only popped up because it instantiated the global launch with launch bounds. The error also happens with a regular `Kokkos::parallel_for(Kokkos::RangePolicy<Kokkos::LaunchBounds<42>>(...), ...)`. It was just a latent bug in some dark untested code path.

For the record, here is the compiler error with `nvcc` 13.1:
```bash
/workspaces/kokkos/core/src/Cuda/Kokkos_Cuda_KernelLaunch.hpp(485): error: more than one instance of overloaded function "Kokkos::Impl::cuda_parallel_launch_global_memory" matches the required type
      return cuda_parallel_launch_global_memory<DriverType, MaxThreadsPerBlock,
             ^
          detected during:
            instantiation of "void *Kokkos::Impl::CudaParallelLaunchKernelFunc<DriverType, Kokkos::LaunchBounds<MaxThreadsPerBlock, MinBlocksPerSM>, Kokkos::Impl::CudaLaunchMechanism::GlobalMemory>::get_kernel_func() [with DriverType=Kokkos::Impl::ParallelFor<Kokkos::Impl::ThenWrapper<<unnamed>::SizedFunctor<Kokkos::View<int, Kokkos::CudaSpace>, 32769UL>>, Kokkos::RangePolicy<Kokkos::Cuda, Kokkos::Impl::IsGraphKernelTag, Kokkos::LaunchBounds<1U, 0U>, void>, Kokkos::Cuda>, MaxThreadsPerBlock=1U, MinBlocksPerSM=0U]" at line 754
            instantiation of "cudaFuncAttributes Kokkos::Impl::CudaParallelLaunchImpl<DriverType, Kokkos::LaunchBounds<MaxThreadsPerBlock, MinBlocksPerSM>, LaunchMechanism>::get_cuda_func_attributes(const Kokkos::Impl::CudaInternal *) [with DriverType=Kokkos::Impl::ParallelFor<Kokkos::Impl::ThenWrapper<<unnamed>::SizedFunctor<Kokkos::View<int, Kokkos::CudaSpace>, 32769UL>>, Kokkos::RangePolicy<Kokkos::Cuda, Kokkos::Impl::IsGraphKernelTag, Kokkos::LaunchBounds<1U, 0U>, void>, Kokkos::Cuda>, MaxThreadsPerBlock=1U, MinBlocksPerSM=0U, LaunchMechanism=Kokkos::Impl::CudaLaunchMechanism::GlobalMemory]" at line 93 of /workspaces/kokkos/core/src/Cuda/Kokkos_Cuda_Parallel_Range.hpp
            instantiation of "void Kokkos::Impl::ParallelFor<FunctorType, Kokkos::RangePolicy<Traits...>, Kokkos::Cuda>::execute() const [with FunctorType=Kokkos::Impl::ThenWrapper<<unnamed>::SizedFunctor<Kokkos::View<int, Kokkos::CudaSpace>, 32769UL>>, Traits=<Kokkos::Cuda, Kokkos::Impl::IsGraphKernelTag, Kokkos::LaunchBounds<1U, 0U>, void>]" at line 125 of /workspaces/kokkos/core/src/Cuda/Kokkos_Cuda_Graph_Impl.hpp
            instantiation of "std::enable_if_t<Kokkos::Impl::is_graph_kernel_v<NodeImpl::kernel_type>, void> Kokkos::Impl::GraphImpl<Kokkos::Cuda>::add_node(std::shared_ptr<NodeImpl>) [with NodeImpl=Kokkos::Impl::GraphNodeImpl<Kokkos::Cuda, Kokkos::Impl::GraphNodeThenImpl<Kokkos::Cuda, Kokkos::Experimental::ThenPolicy<void>, <unnamed>::SizedFunctor<Kokkos::View<int, Kokkos::CudaSpace>, 32769UL>>, Kokkos::Experimental::GraphNodeRef<Kokkos::Cuda, Kokkos::Experimental::GraphNodeRootTag, Kokkos::Experimental::TypeErasedTag>>]" at line 185 of /workspaces/kokkos/core/src/Kokkos_Graph.hpp
            instantiation of "Kokkos::Experimental::Graph<ExecutionSpace> Kokkos::Experimental::create_graph(const Kokkos::Impl::DeviceHandle<ExecutionSpace> &, Closure &&) [with ExecutionSpace=Kokkos::Cuda, Closure=lambda [](const auto &)->auto]" at line 142 of /workspaces/kokkos/core/unit_test/cuda/TestCuda_InterOp_GraphMultiGPU.cpp

1 error detected in the compilation of "/workspaces/kokkos/core/unit_test/cuda/TestCuda_InterOp_GraphMultiGPU.cpp".
gmake[3]: *** [core/unit_test/CMakeFiles/Kokkos_CoreUnitTest_CudaInterOpGraphMultiGPU.dir/build.make:93: core/unit_test/CMakeFiles/Kokkos_CoreUnitTest_CudaInterOpGraphMultiGPU.dir/cuda/TestCuda_InterOp_GraphMultiGPU.cpp.o] Error 1
```

--- a/core/src/Cuda/Kokkos_Cuda_GraphNodeKernel.hpp
+++ b/core/src/Cuda/Kokkos_Cuda_GraphNodeKernel.hpp
@@ -16,7 +16,6 @@
 #include <Kokkos_Parallel_Reduce.hpp>
 
 #include <Cuda/Kokkos_Cuda.hpp>
-#include <Cuda/Kokkos_Cuda_Instance.hpp>
 
 namespace Kokkos {
 namespace Impl {
@@ -135,23 +134,20 @@ class GraphNodeKernelImpl<Kokkos::Cuda, PolicyType, Functor, PatternTag,
   cudaGraphNode_t* get_cuda_graph_node_ptr() const { return m_graph_node_ptr; }
   cudaGraph_t const* get_cuda_graph_ptr() const { return m_graph_ptr; }
 
-  base_t* allocate_driver_memory_buffer(
-      [[maybe_unused]] const CudaInternal& cuda_instance) const {
+  base_t* allocate_driver_memory_buffer() const {
     KOKKOS_EXPECTS(m_driver_storage == nullptr)
 
     const auto& exec = this->get_policy().space();
-
-    KOKKOS_EXPECTS(cuda_instance.m_cudaDev == exec.cuda_device());
-    KOKKOS_EXPECTS(cuda_instance.m_stream == exec.cuda_stream());
 
     std::string alloc_label =
         label + " - GraphNodeKernel global memory functor storage";
     auto mem =
         Kokkos::CudaSpace::impl_create(exec.cuda_device(), exec.cuda_stream());
+    auto* ptr = static_cast<base_t*>(
+        mem.allocate(exec, alloc_label.c_str(), sizeof(base_t)));
     m_driver_storage = std::unique_ptr<base_t, DriverStorageDeleter>(
-        static_cast<base_t*>(
-            mem.allocate(exec, alloc_label.c_str(), sizeof(base_t))),
-        DriverStorageDeleter{.label = alloc_label, .mem = mem});
+        ptr, DriverStorageDeleter{.label = std::move(alloc_label),
+                                  .mem   = std::move(mem)});
     KOKKOS_ENSURES(m_driver_storage != nullptr)
     return m_driver_storage.get();
   }
@@ -178,33 +174,24 @@ struct get_graph_node_kernel_type<KernelType, Kokkos::ParallelReduceTag>
 // <editor-fold desc="get_cuda_graph_*() helper functions"> {{{1
 
 template <class KernelType>
-auto* allocate_driver_storage_for_kernel(const CudaInternal& cuda_instance,
-                                         KernelType const& kernel) {
+auto const& get_graph_node_kernel(KernelType const& kernel) {
   using graph_node_kernel_t =
       typename get_graph_node_kernel_type<KernelType>::type;
-  auto const& kernel_as_graph_kernel =
-      static_cast<graph_node_kernel_t const&>(kernel);
-  return kernel_as_graph_kernel.allocate_driver_memory_buffer(cuda_instance);
+  return static_cast<graph_node_kernel_t const&>(kernel);
 }
 
 template <class KernelType>
 auto const& get_cuda_graph_from_kernel(KernelType const& kernel) {
-  using graph_node_kernel_t =
-      typename get_graph_node_kernel_type<KernelType>::type;
-  auto const& kernel_as_graph_kernel =
-      static_cast<graph_node_kernel_t const&>(kernel);
-  cudaGraph_t const* graph_ptr = kernel_as_graph_kernel.get_cuda_graph_ptr();
+  cudaGraph_t const* graph_ptr =
+      get_graph_node_kernel(kernel).get_cuda_graph_ptr();
   KOKKOS_EXPECTS(graph_ptr != nullptr);
   return *graph_ptr;
 }
 
 template <class KernelType>
 auto& get_cuda_graph_node_from_kernel(KernelType const& kernel) {
-  using graph_node_kernel_t =
-      typename get_graph_node_kernel_type<KernelType>::type;
-  auto const& kernel_as_graph_kernel =
-      static_cast<graph_node_kernel_t const&>(kernel);
-  auto* graph_node_ptr = kernel_as_graph_kernel.get_cuda_graph_node_ptr();
+  auto* graph_node_ptr =
+      get_graph_node_kernel(kernel).get_cuda_graph_node_ptr();
   KOKKOS_EXPECTS(graph_node_ptr != nullptr);
   return *graph_node_ptr;
 }

--- a/core/src/Cuda/Kokkos_Cuda_KernelLaunch.hpp
+++ b/core/src/Cuda/Kokkos_Cuda_KernelLaunch.hpp
@@ -479,7 +479,9 @@ template <class DriverType, unsigned int MaxThreadsPerBlock,
 struct CudaParallelLaunchKernelFunc<
     DriverType, Kokkos::LaunchBounds<MaxThreadsPerBlock, MinBlocksPerSM>,
     CudaLaunchMechanism::GlobalMemory> {
-  static void* get_kernel_func() {
+  static std::decay_t<decltype(cuda_parallel_launch_global_memory<
+                               DriverType, MaxThreadsPerBlock, MinBlocksPerSM>)>
+  get_kernel_func() {
     return cuda_parallel_launch_global_memory<DriverType, MaxThreadsPerBlock,
                                               MinBlocksPerSM>;
   }
@@ -549,7 +551,7 @@ struct CudaParallelLaunchKernelInvoker<DriverType, LaunchBounds,
           cuda_instance, base_t::get_kernel_func(), shmem);
 
       auto* driver_ptr =
-          Impl::allocate_driver_storage_for_kernel(*cuda_instance, driver);
+          get_graph_node_kernel(driver).allocate_driver_memory_buffer();
 
       // Unlike in the non-graph case, we can get away with doing an async copy
       // here because the `DriverType` instance is held in the GraphNodeImpl

--- a/core/src/HIP/Kokkos_HIP_GraphNodeKernel.hpp
+++ b/core/src/HIP/Kokkos_HIP_GraphNodeKernel.hpp
@@ -12,7 +12,6 @@
 #include <Kokkos_Parallel_Reduce.hpp>
 
 #include <HIP/Kokkos_HIP_GraphNode_Impl.hpp>
-#include <HIP/Kokkos_HIP_Instance.hpp>
 
 namespace Kokkos {
 namespace Impl {
@@ -108,23 +107,20 @@ class GraphNodeKernelImpl<Kokkos::HIP, PolicyType, Functor, PatternTag, Args...>
 
   hipGraph_t const* get_hip_graph_ptr() const { return m_graph_ptr; }
 
-  base_t* allocate_driver_memory_buffer(
-      [[maybe_unused]] const HIPInternal& hip_instance) const {
+  base_t* allocate_driver_memory_buffer() const {
     KOKKOS_EXPECTS(m_driver_storage == nullptr);
 
     const auto& exec = this->get_policy().space();
-
-    KOKKOS_EXPECTS(hip_instance.m_hipDev == exec.hip_device());
-    KOKKOS_EXPECTS(hip_instance.m_stream == exec.hip_stream());
 
     std::string alloc_label =
         label + " - GraphNodeKernel global memory functor storage";
     auto mem =
         Kokkos::HIPSpace::impl_create(exec.hip_device(), exec.hip_stream());
+    auto* ptr = static_cast<base_t*>(
+        mem.allocate(exec, alloc_label.c_str(), sizeof(base_t)));
     m_driver_storage = std::unique_ptr<base_t, DriverStorageDeleter>(
-        static_cast<base_t*>(
-            mem.allocate(exec, alloc_label.c_str(), sizeof(base_t))),
-        DriverStorageDeleter{.label = alloc_label, .mem = mem});
+        ptr, DriverStorageDeleter{.label = std::move(alloc_label),
+                                  .mem   = std::move(mem)});
     KOKKOS_ENSURES(m_driver_storage != nullptr);
     return m_driver_storage.get();
   }
@@ -156,22 +152,16 @@ struct get_graph_node_kernel_type<KernelType, Kokkos::ParallelReduceTag>
           Kokkos::ParallelReduceTag>> {};
 
 template <typename KernelType>
-auto* allocate_driver_storage_for_kernel(const HIPInternal& hip_instance,
-                                         KernelType const& kernel) {
+auto const& get_graph_node_kernel(KernelType const& kernel) {
   using graph_node_kernel_t =
       typename get_graph_node_kernel_type<KernelType>::type;
-  auto const& kernel_as_graph_kernel =
-      static_cast<graph_node_kernel_t const&>(kernel);
-  return kernel_as_graph_kernel.allocate_driver_memory_buffer(hip_instance);
+  return static_cast<graph_node_kernel_t const&>(kernel);
 }
 
 template <typename KernelType>
 auto const& get_hip_graph_from_kernel(KernelType const& kernel) {
-  using graph_node_kernel_t =
-      typename get_graph_node_kernel_type<KernelType>::type;
-  auto const& kernel_as_graph_kernel =
-      static_cast<graph_node_kernel_t const&>(kernel);
-  hipGraph_t const* graph_ptr = kernel_as_graph_kernel.get_hip_graph_ptr();
+  hipGraph_t const* graph_ptr =
+      get_graph_node_kernel(kernel).get_hip_graph_ptr();
   KOKKOS_EXPECTS(graph_ptr != nullptr);
 
   return *graph_ptr;
@@ -179,11 +169,7 @@ auto const& get_hip_graph_from_kernel(KernelType const& kernel) {
 
 template <typename KernelType>
 auto& get_hip_graph_node_from_kernel(KernelType const& kernel) {
-  using graph_node_kernel_t =
-      typename get_graph_node_kernel_type<KernelType>::type;
-  auto const& kernel_as_graph_kernel =
-      static_cast<graph_node_kernel_t const&>(kernel);
-  auto* graph_node_ptr = kernel_as_graph_kernel.get_hip_graph_node_ptr();
+  auto* graph_node_ptr = get_graph_node_kernel(kernel).get_hip_graph_node_ptr();
   KOKKOS_EXPECTS(graph_node_ptr != nullptr);
 
   return *graph_node_ptr;

--- a/core/src/HIP/Kokkos_HIP_KernelLaunch.hpp
+++ b/core/src/HIP/Kokkos_HIP_KernelLaunch.hpp
@@ -438,7 +438,7 @@ struct HIPParallelLaunchKernelInvoker<DriverType, LaunchBounds,
 
     if (!Impl::is_empty_launch(grid, block)) {
       auto *driver_ptr =
-          Impl::allocate_driver_storage_for_kernel(*hip_instance, driver);
+          get_graph_node_kernel(driver).allocate_driver_memory_buffer();
 
       // Unlike in the non-graph case, we can get away with doing an async copy
       // here because the `DriverType` instance is held in the GraphNodeImpl

--- a/core/unit_test/cuda/TestCuda_InterOp_GraphMultiGPU.cpp
+++ b/core/unit_test/cuda/TestCuda_InterOp_GraphMultiGPU.cpp
@@ -103,4 +103,53 @@ TEST_F(TEST_CATEGORY_FIXTURE(multi_gpu), then) {
   }
 }
 
+// FIXME Copy/pasted from TestGraph.hpp.
+template <typename ViewType, size_t Count>
+struct SizedFunctor {
+ public:
+  static constexpr size_t count = Count;
+
+  ViewType data;
+
+  SizedFunctor(ViewType data_) : data(std::move(data_)) {}
+
+  KOKKOS_FUNCTION void operator()() const { ++data(); }
+
+ private:
+  std::byte unused[count] = {};
+};
+
+// Create a graph with a then node on the first device that is not the default
+// device. It should exercise the global launch mechanism.
+TEST_F(TEST_CATEGORY_FIXTURE(multi_gpu),
+       then_force_global_launch_on_non_default_device) {
+  using view_t = Kokkos::View<int, Kokkos::CudaSpace>;
+  using functor_t =
+      SizedFunctor<view_t, Kokkos::Impl::CudaTraits::ConstantMemoryUsage + 1>;
+
+  const auto exec_it = std::ranges::find_if(
+      this->execs, [default_device_id = Kokkos::Cuda().cuda_device()](
+                       const Kokkos::Cuda& exec) {
+        return exec.cuda_device() != default_device_id;
+      });
+  ASSERT_NE(exec_it, std::cend(this->execs));
+
+  const view_t data(Kokkos::view_alloc(
+      "data on device " + std::to_string(exec_it->cuda_device()), *exec_it));
+
+  const auto device_handle = Kokkos::Experimental::get_device_handle(*exec_it);
+
+  const auto graph = Kokkos::Experimental::create_graph(
+      device_handle, [&](const auto& root) { root.then(functor_t(data)); });
+
+  graph.submit(*exec_it);
+
+  int value = 0;
+
+  Kokkos::deep_copy(*exec_it, value, data);
+  exec_it->fence();
+
+  ASSERT_EQ(value, 1);
+}
+
 }  // namespace

--- a/core/unit_test/hip/TestHIP_InterOp_GraphMultiGPU.cpp
+++ b/core/unit_test/hip/TestHIP_InterOp_GraphMultiGPU.cpp
@@ -47,16 +47,29 @@ class TEST_CATEGORY_FIXTURE(multi_gpu) : public ::testing::Test {
   std::vector<Kokkos::HIP> execs;
 };
 
-template <typename ViewType>
-struct ThenFunctor {
+// FIXME Copy/pasted from TestGraph.hpp.
+template <typename ViewType, size_t Count>
+struct SizedFunctor {
+ public:
+  static constexpr size_t count = Count;
+
   ViewType data;
 
+  SizedFunctor(ViewType data_) : data(std::move(data_)) {}
+
   KOKKOS_FUNCTION void operator()() const { ++data(); }
+
+ private:
+  std::byte unused[count] = {};
 };
 
-// Create a graph with a then node on any device but the default one.
-TEST_F(TEST_CATEGORY_FIXTURE(multi_gpu), then_on_non_default_device) {
+// Create a graph with a then node on the first device that is not the default
+// device. It should exercise the global launch mechanism.
+TEST_F(TEST_CATEGORY_FIXTURE(multi_gpu),
+       then_force_global_launch_on_non_default_device) {
   using view_t = Kokkos::View<int, Kokkos::HIPSpace>;
+  using functor_t =
+      SizedFunctor<view_t, Kokkos::Impl::HIPTraits::ConstantMemoryUsage + 1>;
 
   const auto exec_it = std::ranges::find_if(
       this->execs, [default_device_id =
@@ -71,8 +84,7 @@ TEST_F(TEST_CATEGORY_FIXTURE(multi_gpu), then_on_non_default_device) {
   const auto device_handle = Kokkos::Experimental::get_device_handle(*exec_it);
 
   const auto graph = Kokkos::Experimental::create_graph(
-      device_handle,
-      [&](const auto& root) { root.then(ThenFunctor<view_t>{.data = data}); });
+      device_handle, [&](const auto& root) { root.then(functor_t(data)); });
 
   graph.submit(*exec_it);
 

